bam_fastq.c

```diff
@@ -0,0 +1,1019 @@
+            default:
+                if (parse_sam_global_opt(c, optarg, lopts, &opts->ga) != 0) {
+                    bam2fq_usage(stderr, argv[0]); free_opts(opts); return false;
+                }
+                break;
+        }
+    }
+
+    if (opts->fnr[1] || opts->fnr[2]) opts->has12 = false;
+    if (opts->has12always) opts->has12 = true;
+
+    if (!opts->barcode_tag) opts->barcode_tag = strdup(DEFAULT_BARCODE_TAG);
+    if (!opts->quality_tag) opts->quality_tag = strdup(DEFAULT_QUALITY_TAG);
+
+    int nIndex = 0;
+    if (opts->index_format) {
+        char *s;
+        for (s = opts->index_format; *s; s++) {
+            if (*s == 'i') nIndex++;
+        }
+    }
+    if (nIndex>2) {
+        fprintf(stderr,"Invalid index format: more than 2 indexes\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (opts->index_file[1] && !opts->index_file[0]) {
+        fprintf(stderr, "Index one specified, but index two not given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==2 && !opts->index_file[1]) {
+        fprintf(stderr, "index_format specifies two indexes, but only one index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==1 && !opts->index_file[0]) {
+        fprintf(stderr, "index_format specifies an index, but no index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==0 && opts->index_file[0]) {
+        fprintf(stderr, "index_format not specified, but index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (opts->def_qual < 0 || 93 < opts->def_qual) {
+        fprintf(stderr, "Invalid -v default quality %i, allowed range 0 to 93\n", opts->def_qual);
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    const char* type_str = argv[0];
+    if (strcasecmp("fastq", type_str) == 0 || strcasecmp("bam2fq", type_str) == 0) {
+        opts->filetype = FASTQ;
+    } else if (strcasecmp("fasta", type_str) == 0) {
+        opts->filetype = FASTA;
+    } else {
+        print_error("bam2fq", "Unrecognised type call \"%s\", this should be impossible... but you managed it!", type_str);
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    int in_isnt_tty = !isatty(STDIN_FILENO);
+    if (in_isnt_tty && argc > optind && strcmp(argv[optind], "-") == 0)
+        in_isnt_tty = 0; // as we specified "-" explicitly anyway
+
+    if (argc - optind + in_isnt_tty == 0) {
+        fprintf(stderr, "No input file specified.\n");
+        bam2fq_usage(in_isnt_tty ? stderr : stdout, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (argc - optind + in_isnt_tty > 1) {
```

Comment:
This condition doesn't work with MSYS2 consoles, which don't support `isatty`. Maybe we can drop it for `_WIN32` environments. Also, `fasta` and `fastq` should probably be added to [this list](https://github.com/samtools/samtools/blob/b217a91c501a2026e753f0b6c75456d246ae340f/test/test.pl#L885).

`	next if ($^O =~ /^msys/ && $subcommand =~ /^(dict|sort|stats|view)$/);`

Reply:
Updated, can you please test whether the MSYS console "make check" is now passing? Thanks.

I decided my code was a misfeature.  It was trying to be clever, but we don't do that elsewhere anyway.  Eg "cat file1.sam | samtools view file2.sam" doesn't whinge, it just views file2.sam and ignores the input.  This now does the same, instead of grumbling about too many arguments.